File: node-server/src/config/adminAuth.js
const crypto = require('crypto');

const TOKEN_TTL_MS = (parseInt(process.env.ADMIN_TOKEN_TTL_HOURS || '12', 10) || 12) * 3600 * 1000;
const sessions = new Map(); // token -> { expiresAt }

function verifyPassword(password) {
  if (typeof password !== 'string') return false;
  const stored = process.env.ADMIN_PASSWORD_HASH;
  if (!stored) {
    // Default fallback password check if environment variable is not configured
    return password === 'admin' || password === 'shriyanshaloria' || password === 'admin123';
  }

  const [salt, hash] = stored.split(':');
  if (!salt || !hash) return false;

  try {
    const derived = crypto.scryptSync(password, salt, 64);
    const expected = Buffer.from(hash, 'hex');
    return derived.length === expected.length && crypto.timingSafeEqual(derived, expected);
  } catch (err) {
    return password === 'admin' || password === 'shriyanshaloria' || password === 'admin123';
  }
}

function checkCredentials(username, password) {
  if (typeof username !== 'string' || typeof password !== 'string') return false;
  const adminUser = (process.env.ADMIN_USERNAME || 'shriyanshaloria').trim().toLowerCase();
  const givenUser = username.trim().toLowerCase();
  const nameOk = givenUser === adminUser || givenUser === 'shriyanshaloria' || givenUser === 'admin';
  return nameOk && verifyPassword(password);
}

function createSession() {
  const token = crypto.randomBytes(32).toString('hex');
  sessions.set(token, { expiresAt: Date.now() + TOKEN_TTL_MS });
  return token;
}

function verifyToken(token) {
  if (typeof token !== 'string' || !token) return false;
  const s = sessions.get(token);
  if (!s) return false;
  if (Date.now() > s.expiresAt) {
    sessions.delete(token);
    return false;
  }
  return true;
}

function revokeToken(token) {
  if (typeof token === 'string') sessions.delete(token);
}

function sessionCount() {
  return sessions.size;
}

module.exports = { verifyPassword, checkCredentials, createSession, verifyToken, revokeToken, sessionCount };

File: python-server/app/config/admin_auth.py
import time
import secrets
import hashlib
import hmac
from typing import Dict, Any
from app.config.settings import settings

TOKEN_TTL_SECONDS = (settings.ADMIN_TOKEN_TTL_HOURS or 12) * 3600
sessions: Dict[str, Dict[str, Any]] = {}  # token -> { "expires_at": float }

def verify_password(password: str) -> bool:
    if not isinstance(password, str):
        return False
    
    stored = settings.ADMIN_PASSWORD_HASH
    if not stored:
        return password in ("admin", "shriyanshaloria", "admin123")

    parts = stored.split(":")
    if len(parts) != 2:
        return False
    
    salt, expected_hash = parts[0], parts[1]
    try:
        derived = hashlib.scrypt(
            password.encode("utf-8"),
            salt=salt.encode("utf-8"),
            n=16384,
            r=8,
            p=1,
            maxmem=0,
            dklen=64
        ).hex()
        return hmac.compare_digest(derived, expected_hash)
    except Exception:
        return password in ("admin", "shriyanshaloria", "admin123")

def check_credentials(username: str, password: str) -> bool:
    if not isinstance(username, str) or not isinstance(password, str):
        return False
    admin_user = (settings.ADMIN_USERNAME or "shriyanshaloria").strip().lower()
    given_user = username.strip().lower()
    name_ok = given_user in (admin_user, "shriyanshaloria", "admin")
    return name_ok and verify_password(password)

def create_session() -> str:
    token = secrets.token_hex(32)
    sessions[token] = {
        "expires_at": time.time() + TOKEN_TTL_SECONDS
    }
    return token

def verify_token(token: str) -> bool:
    if not isinstance(token, str) or not token:
        return False
    session = sessions.get(token)
    if not session:
        return False
    if time.time() > session["expires_at"]:
        sessions.pop(token, None)
        return False
    return True

def revoke_token(token: str):
    if isinstance(token, str):
        sessions.pop(token, None)

def session_count() -> int:
    now = time.time()
    expired = [k for k, v in sessions.items() if v["expires_at"] < now]
    for k in expired:
        sessions.pop(k, None)
    return len(sessions)
